### DISCUSSION TITLE
Examples of using dasel as library in go?

### DISCUSSION THREAD

**Discussion by gozes** [TomWright/dasel#411] [Q&A]
So after having a brief look at the golang docs for dasel as library I can import is not exactly clear to me what the correct/best way to use dasel as library.  the repo dose not have an `example`  folder so  dose anyone knows or has any examples I can look at?

**Comment by TomWright** [comment]
Hey @gozes, there are some basic examples on the [docs site](https://daseldocs.tomwright.me/examples/using-dasel-as-a-go-package).

If these don't help or you have any questions, let me know and I can add more.

**Comment by gozes** [comment]
Will have a better look tomorrow but by just having a quick look now they
are quite basic. I was hoping for example on selecting and filtering  deep
in a nested object :)
.

**Reply by gozes** [reply]
I bit late to come back to this but I did have a better look and yes they are way to basic. It would be really useful to have more complex examples so that one can see how is the library expected to be used

**Reply by TomWright** [reply]
Hey @gozes,

Do you have a description of what you're trying to do and I'll write up something specifically for your use-case and add it to the docs

**Comment by TomWright** [comment]
@gozes Does this example help you understand how to use this from go?

```go
package main

import (
	"fmt"
	"github.com/tomwright/dasel/v2"
)

type NestedValue struct {
	Value any
}

type Root struct {
	Foos  []Foo
	Value NestedValue
}

type Foo struct {
	Bar Bar
}

type Bar struct {
	Baz Baz
}

type Baz struct {
	User User
}

type User struct {
	FirstName string
	LastName  string
}

func main() {
	x := Root{
		Foos: []Foo{
			{
				Bar: Bar{
					Baz: Baz{
						User: User{
							FirstName: "first name 1",
							LastName:  "last name 1",
						},
					},
				},
			},
			{
				Bar: Bar{
					Baz: Baz{
						User: User{
							FirstName: "first name 2",
							LastName:  "last name 2",
						},
					},
				},
			},
		},
		Value: NestedValue{
			Value: NestedValue{
				Value: NestedValue{
					Value: "Hello",
				},
			},
		},
	}

	// Dig into a bunch of nested values
	nestedValues, err := dasel.Select(&x, "Value.Value.Value.Value")
	if err != nil {
		panic(err)
	}
	// Take the final result and cast it to a string
	fmt.Println(nestedValues.Interfaces()[0].(string)) // Hello

	// Read through a slice containing nested values, and take the result as the actual go struct
	users, err := dasel.Select(&x, "Foos.all().Bar.Baz.User")
	if err != nil {
		panic(err)
	}
	for _, u := range users.Interfaces() {
		user := u.(User)
		fmt.Printf("%T, %T, %s, %s\n", u, user, user.FirstName, user.LastName)
		// main.User, main.User, first name 1, last name 1
		// main.User, main.User, first name 2, last name 2
	}
}
```
